[quant][graphmode] Different rule for add/add_/mul/mul_

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38552 [quant] Make MinMaxObserver and MovingAverageMinMaxObserver work with List[Tensor]
* #38436 [quant] Histogram Observer support List[Tensor]
* **#38667 [quant][graphmode] Different rule for add/add_/mul/mul_**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21633555](https://our.internmc.facebook.com/intern/diff/D21633555)

jerryzh168 committed:
[quant][graphmode] Different rule for add/add_/mul/mul_

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1116,10 +1116,12 @@ class TestQuantizeScriptPTSQOps(QuantizationTestCase):
         qconfig_dict = {'': get_default_qconfig(torch.backends.quantized.engine)}
         model = torch.jit.script(module).eval()
         model = quantize_script(model, qconfig_dict, _test_only_eval_fn, [data], inplace=False)
+        if debug:
```
Do you want to have this be part of the final code?

jerryzh168 replied:
yeah I think we can have a debug option

raghuramank100 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -183,7 +179,15 @@ std::vector<std::string> _tensor_info_funcs = {"size"};
 // Aten functions whose output will be quantized or not quantized depending
 // on input tensor
-std::vector<std::string> _propagate_quant_ops = {"cat"};
+std::vector<std::string> _propagate_quant_single_input_ops = {"cat"};
```
NIT: Might be better to say "Input list of tensors"

jerryzh168 replied:
cat actually has one input which is a list of tensor, would like to keep the name since this works for all ops that has single input tensor or input list of tensors.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1267,119 +1269,310 @@ def forward(self, x):
+                         " with instruction set support avx2 or newer.")
     def test_quantized_add(self):
-        class Add(torch.nn.Module):
+        class QuantizedAdd(torch.nn.Module):
```
For brevity, you could reuse all the tests for add for mul, if you make the operator itself a parameter when you define the model.

jerryzh168 replied:
that would be ideal if we can do that... I don't know if it will work with torchscript if we make operator as an argument, but we may have a refactor later

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for add/add_/mul/mul_"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for add/add_/mul/mul_"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21633555](https://our.internmc.facebook.com/intern/diff/D21633555)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for add/add_/mul/mul_"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21633555](https://our.internmc.facebook.com/intern/diff/D21633555)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for add/add_/mul/mul_"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21633555](https://our.internmc.facebook.com/intern/diff/D21633555)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for add/add_/mul/mul_"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21633555](https://our.internmc.facebook.com/intern/diff/D21633555)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for add/add_/mul/mul_"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21633555](https://our.internmc.facebook.com/intern/diff/D21633555)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -674,54 +674,19 @@ static py::dict _jit_debug_module_iterators(Module& module) {
-  "__ixor__",
-  "__str__",
-  "__len__",
+static constexpr const char* magic_method_names[] = {
```
@ZolotukhinM @suo can I commit this?
